File: M3Tools/Controls/People/PhoneNumberField.vb
Imports System.Text.RegularExpressions

Public Class PhoneNumberField
	Public Overrides Property Text As String
		Get
			Return txt_PhoneNumber.Text
		End Get
		Set(value As String)
			txt_PhoneNumber.Text = value
		End Set
	End Property

	Public ReadOnly Property ValidPhone As Boolean
		Get
			Return ValidatePhone()
		End Get
	End Property

	Private Function ValidatePhone() As Boolean
		If Not (String.IsNullOrWhiteSpace(Text) OrElse Regex.IsMatch(Text, "\d{7,10}") OrElse Not txt_PhoneNumber.MaskCompleted) Then
			'  Set error provider for phone number control
			ep_InvalidPhone.SetError(txt_PhoneNumber, "Invalid phone number")
			Return False
		End If

		Return True
	End Function
End Class

File: M3Tools/Controls/People/PhoneNumberField.cs
using System.ComponentModel;
using System.Text.RegularExpressions;

namespace SPPBC.M3Tools
{

    public partial class PhoneNumberField
    {
		/// <summary>
		/// The phone number entered
		/// </summary>
        public string PhoneNumber
        {
            get
            {
                return txt_PhoneNumber.Text;
            }
            set
            {
                txt_PhoneNumber.Text = value;
            }
        }

		/// <summary>
		/// Whether the phone number is valid
		/// </summary>
		[Browsable(false)]
		public bool ValidPhone { get; private set; }

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
        public PhoneNumberField()
        {
            InitializeComponent();

			Validating += new CancelEventHandler(ValidatePhone);
        }

        private void ValidatePhone(object sender, CancelEventArgs e)
        {
            if (!(string.IsNullOrWhiteSpace(Text) || Regex.IsMatch(Text, @"\d{7,10}") || !txt_PhoneNumber.MaskCompleted))
            {
                // Set error provider for phone number control
                ep_InvalidPhone.SetError(txt_PhoneNumber, "Invalid phone number");
                ValidPhone = false;
				e.Cancel = true;
				return;
            }

            ValidPhone = true;
        }
	}
}
